pkg/controller/egress/controller.go

```diff
@@ -128,13 +283,111 @@ func (c *EgressController) processNextEgressGroupWorkItem() bool {
 	return true
 }
 
-func (c *EgressController) syncEgressGroup(key string) error {
+func (c *EgressController) getIPAllocation(egressName string) (net.IP, string, bool) {
+	c.ipAllocationMutex.RLock()
+	defer c.ipAllocationMutex.RUnlock()
+	allocation, exists := c.ipAllocationMap[egressName]
+	if !exists {
+		return nil, "", false
+	}
+	return allocation.ip, allocation.ipPool, true
+}
+
+func (c *EgressController) deleteIPAllocation(egressName string) {
+	c.ipAllocationMutex.Lock()
+	defer c.ipAllocationMutex.Unlock()
+	delete(c.ipAllocationMap, egressName)
+}
+
+func (c *EgressController) setIPAllocation(egressName string, ip net.IP, poolName string) {
+	c.ipAllocationMutex.Lock()
+	defer c.ipAllocationMutex.Unlock()
+	c.ipAllocationMap[egressName] = &ipAllocation{
+		ip:     ip,
+		ipPool: poolName,
+	}
+}
+
+// syncEgressIP is responsible for releasing stale EgressIP and allocating new EgressIP for an Egress if applicable.
+func (c *EgressController) syncEgressIP(egress *egressv1alpha2.Egress) (net.IP, error) {
+	prevIP, prevIPPool, exists := c.getIPAllocation(egress.Name)
+	if exists {
+		// The Egress IP doesn't change, do nothing.
+		if prevIP.String() == egress.Spec.EgressIP && prevIPPool == egress.Spec.ExternalIPPool {
+			return prevIP, nil
+		}
+		// Either EgressIP or ExternalIPPool changes, release the previous one first.
+		c.releaseEgressIP(egress.Name, prevIP, prevIPPool)
+	}
+
+	// Skip allocating EgressIP if ExternalIPPool is not specified and returns whatever user specifies.
```

Comment:
What will happen if user originally specified an Egress IP, and then updated the Spec to add an IPPool, while keeping the Egress IP in the Spec unchanged?
Will we resign the Egress IP to a Node?

Reply:
> What will happen if user originally specified an Egress IP, and then updated the Spec to add an IPPool, while keeping the Egress IP in the Spec unchanged?

Currently controller will only log error that the requested IP is not in the Pool. We could improve it by adding a validationwebhook which rejects the update if the IP doesn't match the pool. 

> Will we resign the Egress IP to a Node?

Good question. Just thought about it after you raised it. I think it will have the problem if we don't add the validationwebhook. Or we have to check if the EgressIP match the Pool in agent too.

Reply:
Added TODO for the validation webhook. Will add it with a separate PR.